### DISCUSSION TITLE
Outdoor Siren Volume controls

### DISCUSSION THREAD

**Discussion by roylofthouse** [tsightler/ring-mqtt#326] [Feature Requests]
First off... Thanks for all the work that everyone has done on this project!

I thought I'd try and help out a little (even if it is for my own benefit) and add support for volume control on the outdoor siren, as I'd like to be able to set it to silent when setting the alarm to "home" and loudest when I'm away via an automation.

I've got this working (of sorts) but need a bit of advice. I've updated siren.js so that there is now a slider that goes from 0-4 (volume control) which sets the five levels in the ring app: silent, very low, low, loud & loudest. The only this is, I've had to remove a couple of checks in the ring-device.js to get it to work.

```
if (isNaN(volume) || volume < 0 || volume > 1) {
            throw new Error('Volume must be between 0 and 1');
        }
        if (!this.supportsVolume) {
            throw new Error(`Volume can only be set on ${ring_types_1.deviceTypesWithVolume.join(', ')}`);
        }

```

The settings for the siren volume are above 1 and the device is not defined as one with volume control. I've looked at adding an entry in ring-types.js but was wondering what's the best way of implementing this? Can we add another device category of of siren and add that to deviceTypesWithVolume?

Also, for some reason, the additional slide entry in lovelace is coming up under it's own entry of "Number" rather than being part of the location:

![image](https://user-images.githubusercontent.com/39500688/155009185-31d09c21-4155-4a79-a935-27a3dc85e214.png)

Code for the control:

```

if (this.device.data.deviceType === 'siren.outdoor-strobe') {
            this.deviceData.mdl = 'Outdoor Siren'
            this.entity.volume = {
                component: 'number',
                min: 0,
                max: 4,
                icon: 'hass:volume-high'
            }
        } else {
            this.deviceData.mdl = 'Siren'
        }        
```

Would be good if I could change the slider for an input select at some point as well.

Any help/advice would be greatly appreciated.

**Comment by roylofthouse** [comment]
So the "Number" thing was just because it wasn't asigned to an area (under advanced). so ignore that one.

**Comment by tsightler** [comment]
It sounds like you are attempting to call setVolume() function from ring-client-api.  I'm not the maintainer of this API library so any changes to that should be directed to that project, but it sounds like it would need to be modifed to work as it actually expects a value between 0-1 (basically, a percentage of volume between 0-100%).  However, you shouldn't require any changes to ring-client-api for this to work, instead just don't bother calling setVolume() as it's just a wrapper function around setInfo() anyway.

If removing the check and calling setVolume() with values of 1-5 actually does successfully change the volume the siren (does this definitely work?) then something like the below should work without removing the check in ring-client-api:
```
const volume = <some_value_between_1_and_5>
device.setInfo({ device: { v1: { volume } } })
```

If the volume is presented in the Ring app as selectable levels, I wonder if it would be better to present as the same list of levels vs using a slider?

Good luck!

**Reply by roylofthouse** [reply]
Thanks for the quick reply.

If I change the the.device.setVolume command to the setInfo one and re-add the two checks in ring-device.js it still works. I can see it changing in the device settings on ring.com when I move the slider.

Is there anyway of passing that without changing the API? I've only ever looked at this stuff today.

Agree that an input selction would be better if it matches the device options.

**Reply by roylofthouse** [reply]
ah, you're meaning the ring-device.js is part of the ring api, so just changing the siren.js would be sufficient? Sorry..been a long day

**Reply by tsightler** [reply]
> Is there anyway of passing that without changing the API? I've only ever looked at this stuff today.

I'm afraid I don't understand the question, if it works with setInfo(), which it should and you indicate it does, then you don't need to change the API, just use setInfo() instead of setVolume() and that's it.  Again, setVolume() is just a wrapper around setInfo() anyway which does nothing more than check if the value is between 0 and 1 before calling setInfo().  Since the siren uses a different set of values for volume, just don't use the built-in setVolume() and call setInfo() directly, no API changes needed at all.  There's a number of cases in the code where I call setInfo() directly for various things where there's no abstraction layer in the API.

I'm also seeing another small artifact with the name as that underscore should not be there, so I need to look at that.

**Reply by tsightler** [reply]
> ah, you're meaning the ring-device.js is part of the ring api, so just changing the siren.js would be sufficient? Sorry..been a long day

Yep, exactly, it's nice that ring-client-api has these helper functions, but there's no requirement to use them, just write a function to set what you want in siren.js via setInfo() and you're good to go.  I use setInfo() in various functions like multi-level-switch (for brightness), and thermostat (for setting temperature), and a few others.

I'd personally suggest using the select component instead, so something like:

```
this.entity.volume = {
    component: 'select',
    options: [ 'Silent', 'Very low', 'Low', 'Loud', 'Loudest' ],
    icon: 'hass:volume-high'
}
```

And then a function to set the select based on number and set the volume based on changes to the selection.  I believe this would more accurately map to the behavior in the Ring app, although I can actually see both sides of this.

**Reply by roylofthouse** [reply]
so I've got both versions working (although I've not got it to select the initial level for the select yet), but I personally prefer the slider (even if it doesn't accurately map the names of the levels.

![image](https://user-images.githubusercontent.com/39500688/155032954-440841c8-edaa-47e1-8229-e320bbc99ec0.png)
![image](https://user-images.githubusercontent.com/39500688/155032876-ed0dc9ab-4848-4dc5-a0a3-d940f7920c35.png)

Is this something you'd be interested in commiting?

**Reply by tsightler** [reply]
Yeah, I think I agree about the slider, it's better to have consistency across volume controls in ring-mqtt than to try to be consistent with the Ring app.  I'm open to merging it since it looks easy, note that any PRs should be sent in against the dev branch.

I do have a question, is the device actually called "Ring_Siren" in the app?  I'm trying to understand why there's an underscore there.  The code should just use the device.name so I feel like the only way that could be happening is of the device name actually includes the underscore.

**Reply by roylofthouse** [reply]
Yes the name matches what is in the ring app. I removed the underscore and it then appeared accordingly in HA, Pull request created (not done one for a while so hope it;s okay).

**Comment by roylofthouse** [comment]
Does the updated version rely on anything from v5? I've run the updated code in my docker image and I'm getting this:

```
2022-02-22T20:02:08.606Z ring-mqtt WARNING - Unhandled Promise Rejection

TypeError [ERR_INVALID_ARG_TYPE]: The "string" argument must be of type string or an instance of Buffer or ArrayBuffer. Received type number (3)

    at new NodeError (node:internal/errors:371:5)

    at Function.byteLength (node:buffer:734:11)

    at publish (/app/ring-mqtt/node_modules/mqtt-packet/writeToStream.js:334:51)

    at Object.generate [as writeToStream] (/app/ring-mqtt/node_modules/mqtt-packet/writeToStream.js:34:14)

    at sendPacket (/app/ring-mqtt/node_modules/mqtt/lib/client.js:176:29)

    at storedPacket (/app/ring-mqtt/node_modules/mqtt/lib/client.js:230:5)

    at Store.put (/app/ring-mqtt/node_modules/mqtt/lib/store.js:42:5)

    at storeAndSend (/app/ring-mqtt/node_modules/mqtt/lib/client.js:225:24)

    at MqttClient._sendPacket (/app/ring-mqtt/node_modules/mqtt/lib/client.js:1238:7)

    at publishProc (/app/ring-mqtt/node_modules/mqtt/lib/client.js:653:14) {

  code: 'ERR_INVALID_ARG_TYPE'

}
```

I'm running 4.9.1, so hoping it just needs something from the 5x release for it to work :-)

**Reply by tsightler** [reply]
If you just dropped the siren.js changes from 5.0 into 4.9.1 then no, it won't work.  In 5.0 I modified the publishMqtt() function to automatically convert values of type number to string using toString() function so I didn't have to keep remembering to convert number values to string when calling it.  Prior versions required using toString() when calling the function, just as you did in your original PR.

If you want to test 5.0 you can just try out the dev branch using he branch feature with your existing Docker image, just set BRANCH=dev as environment for Docker (or set branch config option to "dev" if using the addon) and it will pull in the latest dev branch when you start the container.

**Reply by roylofthouse** [reply]
thought that might be the case. Thanks for all your hard work. It's very much appreciated.

**Reply by roylofthouse** [reply]
So I set branch to dev in HA and although the slider is working it's throwing a warning every time:

![Screenshot_20220222-213450_Home Assistant](https://user-images.githubusercontent.com/39500688/155223661-822b9b6e-ffa3-42ce-8d62-b9aacd5956ac.jpg)

**Reply by tsightler** [reply]
Small bug, should be fixed on a restart.

**Reply by roylofthouse** [reply]
All good after a restart. Thanks again!!

**Reply by bakshiabhi** [reply]
Hello. Can i ask please if this feature exists in 5.5. I cant see or use this. Would be very helpful. Thanks @roylofthouse @tsightler

**Reply by tsightler** [reply]
The code for volume control of the outdoor siren is there and is unchanged from when it was merged back in Feb, so I would assume it would work, but I don't own an Ourdoor Siren and thus have no way to test it and you provided very little detail regarding what you mean by not being able to see or use this.

**Reply by roylofthouse** [reply]
> Hello. Can i ask please if this feature exists in 5.5. I cant see or use this. Would be very helpful. Thanks @roylofthouse @tsightler

Still working fine for me (on the dev branch).

**Reply by bakshiabhi** [reply]
thanks for the quick response. @roylofthouse @tsightler   I can only see these options ( in screenshot) by default on the home assistant gui , i dont see volume up/ down slider or any other options which i can use to change volume from HA. i m running latest HA on raspberry 
![SCR-20221223-nm4](https://user-images.githubusercontent.com/24944891/209373464-c7f92f4d-1306-43ad-b3b6-cbccc494a4b7.png).   do i need to do some thing else to see the volume control, would it be possible to share instructions , thanks

**Reply by roylofthouse** [reply]
Are you using the addon or dev branch in ring-mqtt config? I've just switched back to addon and it becomes unavailable.

**Reply by bakshiabhi** [reply]
i m using addon @roylofthouse , when u say switch on to dev branch , is that easy to do - can u share instructions or link please. Thanks

**Reply by bakshiabhi** [reply]
ah so i can just change branch in config,let  me try, thz

**Reply by roylofthouse** [reply]
https://github.com/tsightler/ring-mqtt/wiki/Installation-(Home-Assistant-Addon)

See branch selection at the bottom of that page

**Reply by tsightler** [reply]
I don't know any reason the addon branch would be different from the dev branch in this regard as there are virtually no changes to anything other than cameras in the dev branch.

On the other hand, what @bakshiabhi post looks like a camera from the native HA ring integration and nothing to do with ring-mqtt at all.

**Reply by roylofthouse** [reply]
For some reason I was still on 4.9.. updated to the latest and the addon branch works..

@bakshiabhi Yes, your controls are not the outdoor siren controls. You should see some that start with "Ring Siren"

**Reply by tsightler** [reply]
> For some reason I was still on 4.9.. updated to the latest and the addon branch works..

Ah, I almost asked if this was somehow the case, but I honestly didn't even think the current dev code would work with <5.x versions of the addon and statistics from HA analytics show that >95% of users are on some version of 5.x.  I guess it shows that my backwards compatibiltiy code that I initially used for migration from 4.x users to 5.x actually works pretty well.

@bakshiabhi I'd strongly recommend that you not use the dev branch as it is in a significant state of flux (I mean, it's dev after all) and should not be required for your use case.  I suppose the next question is, are you seeing any alarm devices at all?  You are specifically asking about the outdoor siren, but that only works with Ring alarm.  Really, probably should start another thread.

**Reply by bakshiabhi** [reply]
i will check how to do this, i clearly have not understood how to use this add on , thanks again
